File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private boolean dfs(TreeNode curr, int min, int max) {
        if (curr == null) {
            return true;
        } else if (curr.val > min && curr.val < max) {
            return dfs(curr.left, min, curr.val) && dfs(curr.right, curr.val, max);
        } else {
            return false;
        }
    }
    public boolean isValidBST(TreeNode root) {
        return dfs(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
}

File: Data Structures & Algorithms/delete-node-in-a-bst/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    TreeNode* searchBST(TreeNode* curr, int key, TreeNode* &prev) {
        while (curr && curr->val != key) {
            prev = curr;

            if (curr->val > key) {
                curr = curr->left;
            } else {
                curr = curr->right;
            }
        }

        return curr;
    }

    TreeNode* findLargestPossibleNode(TreeNode* curr, TreeNode* &prev) {
        while (curr->right) {
            prev = curr;
            curr = curr->right;
        }

        return curr;
    }

    TreeNode* findSmallestPossibleNode(TreeNode* curr, TreeNode* &prev) {
        while (curr->left) {
            prev = curr;
            curr = curr->left;
        }

        return curr;
    }
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (root == nullptr) {
            return nullptr;
        }

        TreeNode* prev = nullptr;
        TreeNode* rm = searchBST(root, key, prev);

        if (rm == nullptr) {
            return root;
        }

        TreeNode* leaf = nullptr;
        if (rm->left) {
            prev = rm;
            leaf = findLargestPossibleNode(rm->left, prev);
        } else if (rm->right) {
            prev = rm;
            leaf = findSmallestPossibleNode(rm->right, prev);
        }

        if (leaf) {
            swap(leaf->val, rm->val);
            rm = leaf;
        }

        if (!prev) {
            delete rm;
            return nullptr;
        }

        TreeNode* attachBranch = (rm->right)? rm->right: rm->left;
        if (prev->right == rm) {
            prev->right = attachBranch;
        } else {
            prev->left = attachBranch;
        }

        delete rm;
        return root;
    }
};